aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -2517,25 +2525,65 @@ void compute_T18_scale_square(
+  // the ramain thing will only be multiply `matrix 0` by (2^4 - 1) times, which
+  // gives us a opportunity to calculate the matrix multiplication in a batch.
+  // The first thing we need to do is sort tensor `s`, which will be helpful to
+  // do the matrix multiplication by range.
+  Tensor sorted_s, sorted_s_inds;
+  // With above example, `sorted_s` is [0, 1, 1, 4], we also will need the index
+  // info, so we can use it to compose the result back.
+  std::tie(sorted_s, sorted_s_inds) = at::sort(s, /*dim=*/0);
+  // Then we call `unique_consecutive` and we will use it to split `sorted_s`,
+  // with above example, `split_counts` is [1, 2, 1].
+  auto split_counts = std::get<2>(at::unique_consecutive(sorted_s, true, /*return_counts=*/true));
+  auto split_counts_cpu = split_counts.to(at::kCPU);
+  // We also need to know the index of the last element of each split, so we can
+  // know how many times we need to do the multiplication for each splited matrix.
+  // Notice that, we will not need to calculate the actual pows, because we will
+  // use the cumulative matrix multiplication.
+  // With about example, `mul_times` will be [0, 1, 3].
+  auto split_edges = at::cumsum(split_counts, /*dim=*/0) - 1;
+  auto unique_s = sorted_s.index_select(0, split_edges).to(at::kLong);
+  auto mul_times = at::diff(unique_s, 1, -1, /*prepend=*/unique_s.new_zeros({1})).to(at::kCPU);
 
   // Square
-  auto mexp_scaled = at::native::compute_T18<scalar_t>(a_scaled);
-  auto s_cpu = (s.device().type() == at::kCPU)
-    ? s : s.to(at::kCPU);
-  for (const auto i : c10::irange(mexp_scaled.size(0))) {
-    auto s_val = s_cpu.select(0, i).template item<int64_t>();
-    auto mexp = mexp_scaled.select(0, i);
-    for (const auto p C10_UNUSED : c10::irange(s_val)) {
-      mexp = at::matmul(mexp, mexp);
+  TORCH_INTERNAL_ASSERT(split_counts_cpu.is_contiguous());
+  TORCH_INTERNAL_ASSERT(mul_times.is_contiguous());
+  auto scs = split_counts_cpu.data_ptr<int64_t>();
+  auto pts = mul_times.data_ptr<int64_t>();
+
+  // We now will do the matrix muplication in a batch, with above example:
+  // 1. Multiply all matrixs by 0 (`mul_times[0]`) times, then do `slice`
+  // to get the remain matrixs by acc[1:] (`split_counts[0]`),
+  // 2. Multiply reamin matrixs by 1 times and slice to acc[2:]
+  // 3. Multiply reamin matrixs by 3 times and slice to acc[1:]
+  // All processed matrices will be stored in `output_pieces`.
+  std::vector<Tensor> output_pieces;
+  auto acc = mexp_scaled.index_select(0, sorted_s_inds);
+  for (int64_t i = 0; i < split_counts_cpu.numel(); ++i) {
+    for (int64_t j = 0; j < pts[i]; j++) {
+        acc = at::matmul(acc, acc);
     }
-    mexp_out.select(0, i).copy_(mexp);
+    output_pieces.push_back(acc.slice(0, 0, scs[i]));
+    acc = acc.slice(0, scs[i]);
   }
+
+  // Compose the result back
+  auto output = at::cat(output_pieces, 0);
+  output = output.index_select(0, at::argsort(sorted_s_inds));
+  mexp_out.copy_(output);
```

Comment:
Passing `mexp_out` does not make sense anymore, as we can simply return `index_select` here rather than performing an unnecessary copy.